### PR TITLE
feat: support external full nodes

### PR DESCRIPTION
**Pull request by reliveyy**
Closes https://github.com/ExchangeUnion/xud-docker/pull/233
Closes https://github.com/ExchangeUnion/xud-docker/issues/216
Closes https://github.com/ExchangeUnion/xud-docker/issues/29

This commit adds configuration files (toml format) to xud-docker. A new container named "exchangeunion/utils" is added to replace most of the bash scripts outside of the docker. So docker and bash will be the only two dependencies after this commit. Configuration files are created automatically when you are going to launch any networks. You can use these files to customize bitcoind, litecoind and geth at the moment.

### CODE REVIEW

**Comment by kilrau** [comment]
Replaces https://github.com/ExchangeUnion/xud-docker/pull/233 ?

**Comment by reliveyy** [comment]
> Replaces #233 ?

Yes.

**Comment by kilrau** [comment]
Other things to be fixed:
- [ ] Update check broken. Asks to update on every run.
- [ ] Update check should show what's being updated, similar to before. Can hide `up-to-date` containers.
- [ ] It is not clear to me how I can check if the config file was successfully read and applied
- [ ] Getting rid of `docker-compose` broke: `start/stop/restart/logs + containername`. We need another way to do this
- [ ] `history`, `network.conf`, `xud-docker.log` are owned by `root`. It has to be current user.
- [ ] In `status` bitcoind shows `created` or `container missing` when using an external bitcoind node via e.g. `mainnet.conf`. It should show the connection status to the external node: either `Connected to remote node` or `Connection to remote node failed`.

I will keep updating this list.

**Comment by kilrau** [comment]
My lndbtc is still not connecting to my external bitcoin node:
`2019-12-16 12:37:29.388 [ERR] LTND: Unable to create chain control: unable to subscribe for zmq block events: dial tcp: lookup bitcoind on 127.0.0.11:53: no such host`

**Comment by reliveyy** [comment]
> My lndbtc is still not connecting to my external bitcoin node

@kilrau I rebuilt and pushed lnd and raiden image on Google Cloud. I think you are still with the old lnd image because I notice I always fail to build lnd on my local machine.

**Comment by erkarl** [comment]
I'm getting the following output when trying to launch testnet:
```
➜  ~ bash xud.sh -b feat/full-node
1) Simnet
2) Testnet
3) Mainnet
Please choose the network: 2
Checking for updates...
Failed to launch testnet environment. For more details, see /home/ar/.xud-docker/testnet/testnet.log
```

/home/ar/.xud-docker/testnet/testnet.log
```
2019-12-17 20:11:30.382 DEBUG 1 --- [ThreadPoolExecutor-0_0] xudenv.testnet: Checking image: exchangeunion/raiden:latest
2019-12-17 20:11:30.383 DEBUG 1 --- [ThreadPoolExecutor-0_1] xudenv.testnet: Checking image: exchangeunion/litecoind:0.17.1
2019-12-17 20:11:30.384 DEBUG 1 --- [ThreadPoolExecutor-0_2] xudenv.testnet: Checking image: exchangeunion/geth:1.9.9
2019-12-17 20:11:30.385 DEBUG 1 --- [ThreadPoolExecutor-0_3] xudenv.testnet: Checking image: exchangeunion/xud:latest
2019-12-17 20:11:30.386 DEBUG 1 --- [ThreadPoolExecutor-0_4] xudenv.testnet: Checking image: exchangeunion/bitcoind:0.18.1
2019-12-17 20:11:30.388 DEBUG 1 --- [ThreadPoolExecutor-0_5] xudenv.testnet: Checking image: exchangeunion/lnd:0.7.1-beta
2019-12-17 20:11:34.112 DEBUG 1 --- [ThreadPoolExecutor-0_3] xudenv.testnet: Image None not found
2019-12-17 20:11:34.776 DEBUG 1 --- [ThreadPoolExecutor-0_4] xudenv.testnet: Image None not found
2019-12-17 20:11:34.846 DEBUG 1 --- [ThreadPoolExecutor-0_1] xudenv.testnet: Image None not found
2019-12-17 20:11:35.370 DEBUG 1 --- [ThreadPoolExecutor-0_2] xudenv.testnet: Image None not found
2019-12-17 20:11:45.200 DEBUG 1 --- [ThreadPoolExecutor-0_1] xudenv.testnet: Comparing images local_image={'sha256': 'dba3ca050122d8877674fca39ff480cc76c9d3611692ea5a103e773e78984c13', 'created': '2019-11-18T09:32:12Z', 'size': 23143082}, remote_image_name='exchangeunion/litecoind:0.17.1', remove_image={'sha256': 'dba3ca050122d8877674fca39ff480cc76c9d3611692ea5a103e773e78984c13', 'created': '2019-11-18T09:32:12Z', 'size': 23143082}
2019-12-17 20:11:45.201 DEBUG 1 --- [MainThread] xudenv.testnet: Checking image: exchangeunion/litecoind:0.17.1: ('up-to-date', 'exchangeunion/litecoind:0.17.1')
2019-12-17 20:11:47.073 DEBUG 1 --- [ThreadPoolExecutor-0_4] xudenv.testnet: Comparing images local_image={'sha256': 'e411e71ae4ba048589477af1085c346a310b2c7e617f1a260ab8c6dd770b7a67', 'created': '2019-10-10T20:07:58Z', 'size': 276170404}, remote_image_name='exchangeunion/bitcoind:0.18.1', remove_image={'sha256': 'e411e71ae4ba048589477af1085c346a310b2c7e617f1a260ab8c6dd770b7a67', 'created': '2019-10-10T20:07:58Z', 'size': 276170404}
2019-12-17 20:11:47.074 DEBUG 1 --- [MainThread] xudenv.testnet: Checking image: exchangeunion/bitcoind:0.18.1: ('up-to-date', 'exchangeunion/bitcoind:0.18.1')
2019-12-17 20:11:55.189 DEBUG 1 --- [ThreadPoolExecutor-0_3] xudenv.testnet: Comparing images local_image={'sha256': 'f092c0795d3635e4bc9b7b4cd6d8833afd1ceb2c79cde01585ab86a5520370de', 'created': '2019-12-12T07:49:48Z', 'size': 439355880}, remote_image_name='exchangeunion/xud:latest', remove_image={'sha256': '8c656a2301bbf1303c961280471e2e0dfffecbd66eba1a7072ddd2eeaa49cc0c', 'created': '2019-12-11T15:23:36Z', 'size': 439355880}
2019-12-17 20:11:55.189 DEBUG 1 --- [MainThread] xudenv.testnet: Checking image: exchangeunion/xud:latest: ('newer', 'exchangeunion/xud:latest')
2019-12-17 20:12:26.735 DEBUG 1 --- [ThreadPoolExecutor-0_5] xudenv.testnet: Comparing images local_image={'sha256': 'c427108de17ca6f7f1291d6ab12deb984e99d2ee37d15443fa4bd177b02e2328', 'created': '2019-12-12T07:49:43Z', 'size': 132915726}, remote_image_name='exchangeunion/lnd:0.7.1-beta__feat-full-node', remove_image={'sha256': '516f3349bf9426815aef74c0a4b09cabf804287266910a749b139c33bd7a4f78', 'created': '2019-12-17T11:59:43Z', 'size': 132891240}
2019-12-17 20:12:28.315 DEBUG 1 --- [ThreadPoolExecutor-0_0] xudenv.testnet: Comparing images local_image={'sha256': '61d4d17d9ebf4fd9c764314371717284f2a0f6e2684dd07c1201df67cffdcf07', 'created': '2019-10-10T20:07:58Z', 'size': 272951959}, remote_image_name='exchangeunion/raiden:latest__feat-full-node', remove_image={'sha256': '768abcb0b1d9ac3150688728d7936c8e018953d991f07b449ec1dd4d1c45ded6', 'created': '2019-12-17T15:07:26Z', 'size': 272977761}
2019-12-17 20:12:28.595 DEBUG 1 --- [ThreadPoolExecutor-0_2] xudenv.testnet: Comparing images local_image={'sha256': '746bfc511275c5756d5122907aaaa3f7e10ab37817882d158f5ecccc3a049720', 'created': '2019-12-06T14:34:00Z', 'size': 45423784}, remote_image_name='exchangeunion/geth:1.9.9', remove_image={'sha256': '16db3708cfafd9eef8ac873f9c285bb1bf0234e870e7508be701875752540747', 'created': '2019-12-12T16:43:15Z', 'size': 45423914}
2019-12-17 20:12:28.596 ERROR 1 --- [MainThread] launcher: Failed to launch
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/launcher/launcher.py", line 542, in launch
    env.start()
  File "/usr/local/lib/python3.8/site-packages/launcher/launcher.py", line 463, in start
    self.check_updates()
  File "/usr/local/lib/python3.8/site-packages/launcher/launcher.py", line 284, in check_updates
    for fut in as_completed(futs, timeout=30):
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 240, in as_completed
    raise TimeoutError(
concurrent.futures._base.TimeoutError: 3 (of 6) futures unfinished
2019-12-17 20:12:28.596 DEBUG 1 --- [MainThread] shell: stop
2019-12-17 20:12:28.597 DEBUG 1 --- [InputHandler] shell.InputHandler: ch='eof', self._cmd=<Command value= index=0 uncommitted=None>
2019-12-17 20:12:28.597 DEBUG 1 --- [InputHandler] shell.InputHandler: End
2019-12-17 20:12:29.362 DEBUG 1 --- [EventLoop] shell.EventLoop: End
```

**Comment by kilrau** [comment]
With go1.13 did:
1. `sudo bash cleanup-processes.sh`
2. `sudo rm -rf go`
3. `sudo rm -rf temp`
4. `npm run test:sim`

```
Target height 1299 reached. Generated 500 blocks.
TokenNetworkRegistry is 0x73c0646E2c9Bf4c4f9Bb237E4Ce40bbc3D7daA97
SecretRegistry is 0x86AA1D25f011B2914daf6Eb546b1fAB8De3d671a
OneToN is 0x240336b00b2ef7FBEAD4108AF045dbf2149FEE73
UserDeposit is 0x59731015eAf394e954C1785119F4393063D9Cd0a
MonitoringService is 0x1390ac3Da8C4571250e169b677203bFd7329b49e
exit status 1
FAIL    github.com/ExchangeUnion/xud-simulation    329.070s
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! xud@1.0.0-mainnet test:sim: `(npm run compile:dev && cd test/simulation && export PATH="$PWD/go/bin:$PATH" && GOPATH=$PWD/go GO111MODULE=on go test -timeout 30m -run=TestIntegration -v)`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the xud@1.0.0-mainnet test:sim script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/kilrau/.npm/_logs/2019-12-18T19_21_14_791Z-debug.log
```

File: images/utils/launcher/shell/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from .shell import Shell
```
Comment (sangaman) [inline images/utils/launcher/shell/__init__.py]: Missing new line.

File: images/utils/launcher/config/xud-docker.conf [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#simnet-dir = "$home_dir/simnet"
+#testnet-dir = "$home_dir/testnet"
+#mainnet-dir = "$home_dir/mainnet"
```
Comment (sangaman) [inline images/utils/launcher/config/xud-docker.conf]: Missing new line.

File: images/utils/launcher/config/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+from .config import Config
```
Comment (sangaman) [inline images/utils/launcher/config/__init__.py]: Missing new line.

**Comment by kilrau** [comment]
I left my env running for 12h+. Geth still consuming tons of CPU+RAM, Raiden still restarting with:
```
Raiden is running in production mode
Checking if the ethereum node is synchronized

You are connected to the 'mainnet' network and the DB path is: /root/.raiden/node_202b9ab1/netid_1/network_d32f5e0f/v23_log.db
Default fee settings are used. If you want use Raiden with mediation fees - flat, proportional and imbalance fees - see https://raiden-network.readthedocs.io/en/latest/overview_and_guide.html#firing-it-up
2019-12-19 14:49:04,542 INFO exited: raiden (exit status 1; not expected)
2019-12-19 14:49:05,546 INFO spawned: 'raiden' with pid 392
2019-12-19 14:49:06,548 INFO success: raiden entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
Welcome to Raiden, version 0.100.5a1.dev161+gf3180af42!
```

**Comment by erkarl** [comment]
Starting a testnet environment got stuck:
```
➜  ~ bash xud.sh -b feat/full-node
1) Simnet
2) Testnet
3) Mainnet
Please choose the network: 2
Checking for updates...
All up-to-date.
^C
➜  ~
➜  ~ docker ps
CONTAINER ID        IMAGE                             COMMAND                  CREATED             STATUS              PORTS                                                  NAMES
72dce86ad48d        exchangeunion/xud:latest          "/usr/bin/supervisor…"   18 minutes ago      Up 18 minutes       8885/tcp, 8887/tcp, 0.0.0.0:18885->18885/tcp           testnet_xud_1
43920efa89e4        exchangeunion/raiden:latest       "/usr/bin/supervisor…"   18 minutes ago      Up 18 minutes       5001/tcp                                               testnet_raiden_1
e138f7ad596d        exchangeunion/geth:1.9.9          "./entrypoint.sh"        18 minutes ago      Up 8 minutes        8545-8547/tcp, 30303/tcp, 30303/udp                    testnet_geth_1
350d68946b19        exchangeunion/lnd:0.8.2-beta      "/usr/bin/supervisor…"   18 minutes ago      Up 18 minutes       9735/tcp, 10009/tcp, 10735/tcp, 19735/tcp, 20735/tcp   testnet_lndltc_1
637e55426d2b        exchangeunion/lnd:0.8.2-beta      "/usr/bin/supervisor…"   18 minutes ago      Up 18 minutes       9735/tcp, 10009/tcp, 10735/tcp, 19735/tcp, 20735/tcp   testnet_lndbtc_1
25e4b1fb8680        exchangeunion/litecoind:0.17.1    "/entrypoint.sh"         18 minutes ago      Up 8 minutes        9332/tcp, 19332/tcp, 29332-29333/tcp                   testnet_litecoind_1
cb0ec54c7cfa        exchangeunion/bitcoind:0.19.0.1   "bitcoind -server -r…"   18 minutes ago      Up 8 minutes        8332/tcp, 18332/tcp, 28332-28333/tcp                   testnet_bitcoind_1
```

After 15 minutes I stopped the bash script, but looks like the containers are running.

**Comment by erkarl** [comment]
> Starting a testnet environment got stuck:
> 
> ```
> ➜  ~ bash xud.sh -b feat/full-node
> 1) Simnet
> 2) Testnet
> 3) Mainnet
> Please choose the network: 2
> Checking for updates...
> All up-to-date.
> ^C
> ➜  ~
> ➜  ~ docker ps
> CONTAINER ID        IMAGE                             COMMAND                  CREATED             STATUS              PORTS                                                  NAMES
> 72dce86ad48d        exchangeunion/xud:latest          "/usr/bin/supervisor…"   18 minutes ago      Up 18 minutes       8885/tcp, 8887/tcp, 0.0.0.0:18885->18885/tcp           testnet_xud_1
> 43920efa89e4        exchangeunion/raiden:latest       "/usr/bin/supervisor…"   18 minutes ago      Up 18 minutes       5001/tcp                                               testnet_raiden_1
> e138f7ad596d        exchangeunion/geth:1.9.9          "./entrypoint.sh"        18 minutes ago      Up 8 minutes        8545-8547/tcp, 30303/tcp, 30303/udp                    testnet_geth_1
> 350d68946b19        exchangeunion/lnd:0.8.2-beta      "/usr/bin/supervisor…"   18 minutes ago      Up 18 minutes       9735/tcp, 10009/tcp, 10735/tcp, 19735/tcp, 20735/tcp   testnet_lndltc_1
> 637e55426d2b        exchangeunion/lnd:0.8.2-beta      "/usr/bin/supervisor…"   18 minutes ago      Up 18 minutes       9735/tcp, 10009/tcp, 10735/tcp, 19735/tcp, 20735/tcp   testnet_lndbtc_1
> 25e4b1fb8680        exchangeunion/litecoind:0.17.1    "/entrypoint.sh"         18 minutes ago      Up 8 minutes        9332/tcp, 19332/tcp, 29332-29333/tcp                   testnet_litecoind_1
> cb0ec54c7cfa        exchangeunion/bitcoind:0.19.0.1   "bitcoind -server -r…"   18 minutes ago      Up 8 minutes        8332/tcp, 18332/tcp, 28332-28333/tcp                   testnet_bitcoind_1
> ```
> 
> After 15 minutes I stopped the bash script, but looks like the containers are running.

Can't reproduce this. Trying again on a fresh environment worked.

**Comment by reliveyy** [comment]
@kilrau You can post xud-docker.log, mainnet.log, testnet.log, simnet.log to provide more details about your environment.

**Comment by kilrau** [comment]
config files missing
```
#infura-project-id = ""
#infura-project-secret = ""
```

**Comment by reliveyy** [comment]
@kilrau If you already have your config file then new config file will not copied to replace your old one.

**Comment by kilrau** [comment]
I think providing config files as `sample-mainnet.conf` and the user needs to `cp sample-mainnet.conf mainnet.conf` is the better option and industry standard. The only way to update config file options.

File: images/utils/launcher/shell/__init__.py (hunk 1 above)
Comment (reliveyy) [reply images/utils/launcher/shell/__init__.py]: @sangaman Fixed.

File: images/utils/launcher/config/xud-docker.conf (hunk 2 above)
Comment (reliveyy) [reply images/utils/launcher/config/xud-docker.conf]: @sangaman Fixed.

File: images/utils/launcher/config/__init__.py (hunk 3 above)
Comment (reliveyy) [reply images/utils/launcher/config/__init__.py]: @sangaman Fixed. Thanks for pointing out the formatting issue.

**Comment by kilrau** [comment]
Did some more testing, updated issue list:
- [ ] After testing infura (works!), switching `mainnet.conf` back to `geth` `external = false`, raiden complains about and exits (never saw this before):
```
The underlying ethereum node does not have the web3 rpc interface enabled. Please run it with --rpcapi eth,net,web3,txpool for geth and --jsonrpc-apis=eth,net,web3,parity for parity.
2019-12-28 10:12:30,146 INFO exited: raiden (exit status 1; not expected)
```
- [ ] In this case: https://paste.ubuntu.com/p/dkgFCBpbMV/ when raiden logs print `Waiting for the ethereum node to synchronize.`, `status` for raiden should show `Waiting for sync` too, like the other L2 clients.
- [ ] xud can't connect to raiden (`xucli getinfo` shows `connect ECONNREFUSED 10.0.3.7:5001`). Update: we found that raiden for some mysterious reason is not listeing on 5001. Update02: it looks like Raiden is running fine for some seconds, then crashes and gets restarted by supervisord. This in infinity loop. See https://github.com/ExchangeUnion/xud-docker/pull/246#issuecomment-567827784
```
kilrau@K-Yoga:~/.xud-docker/mainnet$ docker exec -it mainnet_raiden_1 bash
bash-5.0# netstat -ant | grep LISTEN
tcp        0      0 127.0.0.11:42425        0.0.0.0:*               LISTEN 
```
- In this case where some images are outdated it should ask me for permission to update (it just did update without asking for Y/n:
```
kilrau@K-Yoga:~/.xud-docker/mainnet$ xud -b feat/full-node
1) Simnet
2) Testnet
3) Mainnet
Please choose the network: 3
Checking for updates...
* Image exchangeunion/raiden:latest: outdated
* Image exchangeunion/geth:1.9.9: outdated
* Container mainnet_bitcoind_1: missing
* Container mainnet_litecoind_1: missing
* Container mainnet_geth_1: missing
* Container mainnet_lndbtc_1: missing
* Container mainnet_lndltc_1: missing
* Container mainnet_raiden_1: missing
* Container mainnet_xud_1: missing
Pulling exchangeunion/raiden:latest__feat-full-node
Retagging exchangeunion/raiden:latest__feat-full-node to exchangeunion/raiden:latest
Pulling exchangeunion/geth:1.9.9__feat-full-node
Retagging exchangeunion/geth:1.9.9__feat-full-node to exchangeunion/geth:1.9.9
Creating mainnet_bitcoind_1
Creating mainnet_litecoind_1
Creating mainnet_geth_1
Creating mainnet_lndbtc_1
Creating mainnet_lndltc_1
Creating mainnet_raiden_1
Creating mainnet_xud_1
```

**Comment by reliveyy** [comment]
@kilrau Choosing network process is handled in utils container and Docker doesn't provided a API to modify the container's name after running. So it's impossible to change utils container's name to <network>_utils_1. But I still changed the random name to utils_<timestamp> for better readability.

**Comment by kilrau** [comment]
> @kilrau Choosing network process is handled in utils container and Docker doesn't provided a API to modify the container's name after running. So it's impossible to change utils container's name to <network>_utils_1 unless we move choosing network logic out of utils container.

Let's move it out.

**Comment by kilrau** [comment]
Test cases as discussed:
* delete "home-dir" entirely, specifying via cli parameter or .xud-docker.conf should result in an parser error (print this error `Couldn't recognize paramter "home-dir", exiting...`) and should exit the script. All parameter parsing errors should print the parsing error exit the script, DO NOT RUN.
* if I set `#mainnet-dir = "/media/HDD/"` and there is no directory yet (HDD is empty), it should ask   `/media/HDD/data` all other mainnet files like `mainnet.conf` are in `/media/HDD`.
* if I set `#mainnet-dir = "/media/HDD/xud"` and there is no `xud` directory yet (HDD is empty), it should ask if I want to create  `/media/HDD/xud` data folder is in `/media/HDD/data`.
* if specified directory is not available when launching xud, print sth like `$network-dir not available, please check and run again.`
* only create `simnet` , `testnet` , `mainnet` subfolders if user chooses this network
* if I run xud.sh with `~/.xud-docker/xud-docker.conf` present specifiying another folder, there shouldn't be any $network subfolders created in `~/.xud-docker/`.
* no `_1` addon on container names, utils container gets created after user selects network as e.g. `mainnet_utils`

**Comment by reliveyy** [comment]
> delete "home-dir" entirely, specifying via cli parameter or .xud-docker.conf should result in an parser error (print this error Couldn't recognize paramter "home-dir", exiting...) and should exit the script. All parameter parsing errors should print the parsing error exit the script, DO NOT RUN.

Done. But with different error message like `error: unrecognized arguments: --home-dir`

> if I set #mainnet-dir = "/media/HDD/" and there is no directory yet (HDD is empty), it should ask /media/HDD/data all other mainnet files like mainnet.conf are in /media/HDD.
> if I set #mainnet-dir = "/media/HDD/xud" and there is no xud directory yet (HDD is empty), it should ask if I want to create /media/HDD/xud data folder is in /media/HDD/data.

Done. It is implemented with a question: `Would you like to create directory: $NETWORK_DIR? [Y/n]`
* The 1st case will have no question because HDD is empty but it exists
* The 2nd case will have a question: `Would you like to create directory: /media/HDD/xud`

> if specified directory is not available when launching xud, print sth like $network-dir not available, please check and run again.

Done. It is implemented by check if `$NETWORK_DIR` is a directory, readable and writable.

> only create simnet , testnet , mainnet subfolders if user chooses this network. Only create directory of chosen network.
> if I run xud.sh with ~/.xud-docker/xud-docker.conf present specifiying another external folder, there shouldn't be any $network subfolders created in ~/.xud-docker/.

Done.

> no _1 addon on container names, utils container gets created after user selects network as e.g. mainnet_utils

WIP. Need to make this change backward compatible.

**Comment by erkarl** [comment]
I'm getting the following output when `~/.xud-docker` does not exist:
```
➜  ~ bash xud.sh -b feat/full-node
1) Simnet
2) Testnet
3) Mainnet
Please choose the network: 3
🚀 Launching mainnet environment
Would you like to create directory: /home/ar/.xud-docker/mainnet? [Y/n] Y
mkdir: cannot create directory ‘/home/ar/.xud-docker/mainnet’: Permission denied
```

**Comment by erkarl** [comment]
Let's also squash the commits once all the fixes are in.

**Pull request merged**